### DISCUSSION TITLE
Does it support solving electrophysiological equations on surface meshes?

### DISCUSSION THREAD

**Discussion by wufanxiaozi** [Chaste/Chaste#329] [Q&A] [answered]
I would like to know if the electrophysiological equations support curved meshes? I saw that it is possible to create 2D cells when the mesh reads, 3D space. 
```c++
            TrianglesMeshReader<2,3> mesh_reader("mesh/test/data/disk_in_3d");
            TetrahedralMesh<2,3> mesh;
            mesh.ConstructFromMeshReader(mesh_reader);
            c_vector <double, 2> edge_len = mesh.CalculateMinMaxEdgeLengths();
            TS_ASSERT_DELTA(edge_len[0], 0.0628, 1e-4);
            TS_ASSERT_DELTA(edge_len[1], 0.2010, 1e-4);
``` 
But when I create a single domain problem with 2D unit 3D space, I get an error. I see that there is currently no option for 2D unit 3D space.
```c++
// Monodomain
template class AbstractCardiacProblem<1, 1, 1>;
template class AbstractCardiacProblem<1, 2, 1>;
template class AbstractCardiacProblem<1, 3, 1>;
template class AbstractCardiacProblem<2, 2, 1>;
template class AbstractCardiacProblem<3, 3, 1>;

// Bidomain
template class AbstractCardiacProblem<1, 1, 2>;
template class AbstractCardiacProblem<2, 2, 2>;
template class AbstractCardiacProblem<3, 3, 2>;

// Extended Bidomain
template class AbstractCardiacProblem<1, 1, 3>;
template class AbstractCardiacProblem<2, 2, 3>;
template class AbstractCardiacProblem<3, 3, 3>;
``` 
Has anyone tried electrophysiological solutions for surface meshes? I would be grateful if someone could answer this question.

**Comment by jmpf** [comment]
I don't think that anyone has.  The fact that the templates are not instantiated in the main trunk means that we are not currently testing an implementation.  So you could try implementing a solver since the code is templated and might work out of the box.  (I would be cautious about how to apply the correct boundary conditions though.)

The developer who wrote most of the Extended Bidomain code did so for gastric modelling and possibly was considering surface meshes.  This was more than 10 years ago and I may be misremembering.  @mirams Do you know of anyone using surface meshes or remember if Alberto made an implementation?

**Reply by mirams** [reply]
It appears that the Auckland team have just done some gastric electrophys simulations on 2D in 3D space as well as 3D simulations with Chaste in this paper: https://www.mdpi.com/2076-3417/14/22/10393 it might be worth getting in touch with them to see how they did it!

**Comment by OmkarAthavale** [comment, marked as answer]
Hello, I am one of the Auckland team and I used Chaste for simulating gastric electrophysiology on a 2D shell mesh in 3D space. As @jmpf wrote, I just instantiated a few classes with the correct dimensions. Here is the [commit diff](https://github.com/user-attachments/files/18482773/commit-01a17ec.txt) for this change (parent commit is: [a18592](https://github.com/Chaste/Chaste/commit/a18592)). 

For a Laplace problem on a stomach shell mesh we tested applying Dirichlet boundary conditions and the solution was as expected (we never needed to apply Neumann conditions for our work so have not tested that):
```
Loop through boundary nodes with TetrahedralMesh<2,3>::BoundaryNodeIterator 
    if node matches boundary location within some tolerance
        add node to BoundaryConditionsContainer<2,3,1> with an appropriate ConstBoundaryCondition<3>*
```

Hope that helps!

**Reply by mirams** [reply]
That is fantastic, thanks @OmkarAthavale, if it's that simple we should probably instantiate and test that in the main codebase. I've opened [a ticket](https://github.com/Chaste/Chaste/issues/359) to investigate.
